functions/find_first.yml

```diff
@@ -0,0 +1,64 @@
+name: find_first
+topic: strings
+args:
+  required:
+  - name: subject
+    type: [string]
+    desc: 'Subject string'
+  - name: sub
+    type: [string]
+    desc: 'Substring'
+  optional: 
+    - name: start
+      type: [number]
+      desc: 'Position in the subject string where searching should start.'
+    - name: end
+      type: [number]
+      desc: 'Position in the subject string where searching should end.'
+returns:
+  type: number
+  desc: ''
+desc: |
+  Given the `subject` string, `find_first()` returns the index of the first occurence where the `sub` substring appears in `subject` or `null`.
+
+  The `start` and `end` parameters are optional and allow to select where `find_first()` must perform its search within `subject`.
+
+  * If `start` is omitted, it defaults to `0` which is the start of the `subject` string.
+  * If `end` is omitted, it defaults to `length(subject) - 1` which is is the end of the `subject` string.
+examples:
+  find_first:
+    context: &subject subject string
+    args: ['@', 'string']
+    returns: 8.0
+  find_first_start:
+    context: *subject
+    args: ['@', 'string', '`0`']
+    returns: 8.0
+  find_first_start_end:
+    context: *subject
+    args: ['@', 'string', '`0`', '`14`']
+    returns: 8.0
+  find_first_start_before_end_after:
+    context: *subject
+    args: ['@', 'string', '`-99`', '`100`']
```

Comment:
Well, in my view, `count` does support out of bounds values, in that you can specify more replacements than would be possible. However, you can not specify negative values because that would not be consistent for a count.

Basically, `replace()` and `split()` expect a _count_ number of replacements. Cannot be negative, but can be unbounded.
`find()` expects a range expressed using _offsets_ which can be negative. A bit like slice.

Reply:
> Well, in my view, `count` does support out of bounds values, in that you can specify more replacements than would be possible. However, you can not specify negative values because that would not be consistent for a count.
> 
> Basically, `replace()` and `split()` expect a _count_ number of replacements. Cannot be negative, but can be unbounded. `find()` expects a range expressed using _offsets_ which can be negative.

Given that ```replace($subject, $old, $new, `0`)``` is explicitly valid and interpreted as "_perform up to 0 replacements_" (i.e., do not change $subject), how would interpreting ```replace($subject, $old, $new, `-1`)``` in the same way (i.e., "_perform up to −1 replacements_") be inconsistent[^1]? I raised this point because it _does_ seem inconsistent to require a non-negative count for `replace`/`split`/etc. but not to require a non-negative index for `find` functions (**especially** since there is reasonable support for an expectation that a negative index would be interpreted relative to the end, which now that I think of it might be better behavior[^2]).

> A bit like slice.

Slices interpret negative indices as relative to the _end_: https://jmespath.site/main/#spec-slices

[^1]: e.g., Python `"foo".replace("o", "~", -1)` [performs the first −1 replacements](https://docs.python.org/3/library/stdtypes.html#str.replace) (i.e., none) and successfully returns `"foo"` unmodified.
[^2]: e.g., Python `"foo bar baz".find("a", -3)` returns `9`, the index of the _second_ "a", because the position arguments "[_are interpreted as in slice notation_](https://docs.python.org/3/library/stdtypes.html#str.find)" and the first "a" is not found at or after index 11 - 3 = 8. In contrast, though, JavaScript `"foo bar baz".indexOf("a", -3)` [clamps the position argument to be nonnegative](https://tc39.es/ecma262/multipage/text-processing.html#sec-string.prototype.indexof) and returns 5.

Reply:
Covered in #119: https://github.com/jmespath-community/jmespath.spec/pull/119#discussion_r1008717440

Reply:
As you’ve outlined, `find_first()` and  `find_last()` accept offsets that can indeed be interpreted in slice notation. That is one of the clarifications I made explicit.

I understand we could have the `$count` parameter be interpreted more like a `$pos` offset starting at `0` to represent the end of the string. However, I’m not happy to go down that path.

I think the core of our argument here is that `$count` can be `0`. I would not be completely against enforcing `$count` to start from `1` and forbid the value `0` as I do not think this would actually limit the usefulness of the `split()` and `replace()` functions. Even though I do not see any harm in accepting a `$count` of `0`.

What do you think?

Reply:
I think what's important to get right is consistency across functions and syntax with respect to how out-of-bounds input (and unusual input in general) are treated, to avoid frustration about a pattern working in one context but not in another. Some languages are liberal while others are strict, and both are fine positions but a mixture really isn't.

So for example, if `'foo'[-100:100]` is valid and results in "foo" (i.e., slice indices are clamped to the valid range), then an "up to _N_" $count should accept negative values and clamp them to the valid range and the overall experience will align pretty well with Python. There's also a consideration of what to do with non-integer input where valid behavior requires interpretation as an integer, which doesn't come up at all in Python because (unlike in JSON and JMESPath) there is an integer type.

Some relevant ambiguities in JMESPath documentation:
* out-of-bounds string indices, e.g. `'foo'[10]` and `'bar'[-10]` (noting that an out-of-bounds array index such as `[][1]` returns `null` per [Index Expressions](https://jmespath.site/#spec-index-expressions))
* out-of-bounds slice indices, e.g. ``` `[]`[10:] ``` and `'foo'[-100:100]`
* `contains(string, nonString)`, e.g. ```contains('misconstrue', `true`)``` and ```contains('blue42', `42`)``` and ```contains('blue42', `42.0`)``` and ```contains('foo', `["foo"]`)``` and ```contains('[0,1,2]', `[0,1,2]`)``` (noting that non-overloaded functions such as `ends_with` and `join` reject non-string input where a string is expected)
* `find` function with non-integer position, e.g. ```find_first('foo', 'o', `1.1`)```
* `from_items(invalidArray)`, e.g. ```from_items(`[ [] ]`)``` and ```from_items(`[ ["keyWithNoValue"] ]`)``` and ```from_items(`[ [1, "valueForNonstringKey"], [null, "valueForNonstringKey"], [{}, "valueForNonstringKey"] ]`)```
* `pad` function with non-integer length, e.g. ```pad_left('foo', `4.2`)```
* `replace`/`split` with non-integer count, e.g. ```split('foo bar baz', ' ', `1.3`)```

Each of those is an issue unto itself, but the resolution of them should be coordinated holistically.

Reply:
Thanks for pointing out that negative offsets are indeed valid for `index-expression`.  I actually had forgotten about this peculiarity.

Let‘s divide your points above.

# contains function

The `contains()` function is overloaded to accepting either an array or a string. But I do not see any inconsistency in the specification. What you are observing is that the JavaScript implementation still supports old-style literal parsing strings that have been deprecated in JEP-12. This means that `` `foo` `` is no longer a valid standards-compliant string.

In fact the JMESPath.Net implementation (that you can test at https://jmespath.site/#preview) does not accept that as a valid string. By looking into this, I noticed that the .Net implementation seems guilty as well to coerce `` `42` `` and `` 42.0 `` to a valid string. 😒

1 `` contains('misconstrue', `true`) `` means: does the string `"misconstrue"` contain the boolean `true`?
2 `` contains('blue42`, `42` ``: does `"blue42"` contain the number `42`?
3 `` contains('blue42`, `42.0` ``: does `"blue42"` contain the number `42.0`?
4 `` contains('foo', `["foo"]`) ``: does `"foo"` contain the JSON array `[ "foo" ]`?
5 `` contains('[0,1,2]', `[0,1,2]`) ``: does the `"[0,1,2]"` contains the JSON array `[0, 1, 2]`?

Again, the JavaScript implementation does indeed accept those examples and anwsers `true` for expressions 1, 2 and 3.
The fact that it answers `true` for expressions 4 and 5 baffles me. But that does not bother me too much. Nowadays, I’m looking at the Python implementation that fails with a syntax error as per the specification.

In the end, all those expressions are rejected by the existing specification.
Our improvements are designed to build upon this existing specification.

# Non-integer arguments

The `$count` argument to various functions is new to all our proposal. Do you qualify as an inconsistency the fact that numbers are accepted as valid only if they are integers? Do you feel we should accept numbers and truncate them to their integral parts ? That’s something I would be OK with.

# Negative offsets interpreted as slice notation… but not always

It seems the point you are trying to make is that the specification does allow negative offsets in `index-expression` even though those offsets do not work like slice in that it does not always end up referring to a valid index in the array. That is something I can certainly agree. 

So we would probably need a JEP at some point to fix this inconsistency.

However, in my mind, an _array index_ is still different than a `count` number.

Reply:
> # contains function
> The `contains()` function is overloaded to accepting either an array or a string. But I do not see any inconsistency in the specification. What you are observing is that the JavaScript implementation still supports old-style literal parsing strings that have been deprecated in JEP-12. This means that `` `foo` `` is no longer a valid standards-compliant string.

I think you're misunderstanding a bit... the consistency of `contains(string, nonString)` can't even be evaluated because the spec itself does not define _any_ behavior for such invocations. For a string to contain a non-string isn't a meaningful concept without some text explaining it (e.g., "_this function returns true if the string contains **the result of `to_string` on** the provided <code>$search</code> argument_").

<blockquote><p dir="auto">1 <code class="notranslate">contains('misconstrue', `true`)</code> means: does the string <code class="notranslate">"misconstrue"</code> contain the boolean <code class="notranslate">true</code>?<br>
2 <code class="notranslate">contains('blue42`, `42`</code>: does <code class="notranslate">"blue42"</code> contain the number <code class="notranslate">42</code>?<br>
3 <code class="notranslate">contains('blue42`, `42.0`</code>: does <code class="notranslate">"blue42"</code> contain the number <code class="notranslate">42.0</code>?<br>
4 <code class="notranslate">contains('foo', `["foo"]`)</code>: does <code class="notranslate">"foo"</code> contain the JSON array <code class="notranslate">[ "foo" ]</code>?<br>
5 <code class="notranslate">contains('[0,1,2]', `[0,1,2]`)</code>: does the <code class="notranslate">"[0,1,2]"</code> contains the JSON array <code class="notranslate">[0, 1, 2]</code>?</p>
<p dir="auto"></p></blockquote>

> Nowadays, I’m looking at the Python implementation that fails with a syntax error as per the specification.

All of the above examples are _syntactically_ valid; any implementation that fails to parse them would be broken. But there is no spec-defined runtime behavior for their _evaluation_, which is a problem to correct.

> In the end, all those expressions are rejected by the existing specification.

Can you point out what part of the specification defines such rejection? I don't see anything of the sort at https://jmespath.site/main/#func-contains :
<blockquote>
boolean contains(array|string $subject, any $search)
<p>Returns <code>true</code> if the given <code>$subject</code> contains the provided <code>$search</code>
string.</p><p>If <code>$subject</code> is an array, this function returns true if one of the elements
in the array is equal to the provided <code>$search</code> value.</p><p>If the provided <code>$subject</code> is a string, this function returns true if the string contains the provided <code>$search</code> argument.</p></blockquote>

But for the record, I am very much in favor of `contains` rejecting any non-string $search with a string $subject, which would feel most consistent with the otherwise good type checking of functions in general and string functions in particular.

> # Non-integer arguments
> The `$count` argument to various functions is new to all our proposal. Do you qualify as an inconsistency the fact that numbers are accepted as valid only if they are integers? Do you feel we should accept numbers and truncate them to their integral parts ? That’s something I would be OK with.

There is no specification text that I can see requiring $count arguments to be integers, but I would be in favor of defining such behavior (effectively carving out an "integer" subtype).

> # Negative offsets interpreted as slice notation… but not always
> It seems the point you are trying to make is that the specification does allow negative offsets in `index-expression` even though those offsets do not work like slice in that it does not always end up referring to a valid index in the array. That is something I can certainly agree.
> 
> So we would probably need a JEP at some point to fix this inconsistency.

I can imagine arguments for `'foo'[-100:100]` to either result in an error or for it to be equivalent to `'foo'`, and I'm not sure there's sufficient precedent within JMESPath to provide sufficient support for one vs. the other. What I _can't_ imagine is an argument for leaving the behavior undefined, as is currently the case AFAICT.

> However, in my mind, an _array index_ is still different than a `count` number.

I agree, but it would feel strange to me if `'foo'[100]` and `'foo'[-100:100]` are valid but ```split('foo bar baz', ' ', -1)``` raises an error, because the first two results are consistent with a language that is generally liberal with respect to numbers but the last is not.

Reply:
I think I understand your point and am completely aligned with respect to always be explicit and try to never leave behaviour undefined. For what I can control, I will make sure to follow this trend.

I have actually [updated the JEP-14]( https://github.com/jmespath-community/jmespath.spec/blob/main/jep-014-string-functions.md) to require integers for the `$count` parameter to the `split()` and `replace()` functions. Some of this text has no yet made it to the spec but that is in my to-do list. Of course, I will address the empty string corner case as well.

As for the ` find_first()` and `find_last()` functions I would be OK'ish to allow negative values as seems to be the case for `index-expression`. But in that case, they would not follow slice conventions either, as `index-expression` indices currently do not... Which is a bit silly I think. I am weary to standardize a quirky behaviour, though, and would be more keen to deprecate and forbid negative indices for `index-expression` in the future instead. Which means as imperfect as it is, I would still leave negative indices for `find_` functions currently undefined for now.

Lastly, `contains()` seems to have a well defined behavior when its first argument is an array, as it is based upon equality with one of the array element and equality is well defined in JMESPath. You are absolutely right, `contains` is not sufficiently defined for string arguments. I think a new JEP is warranted to close the loop and tighten its behavior to only accept a string `search` string and define _containment_ by the intuitive occurence of an _equal_ substring within the subject string.

Reply:
> As for the ` find_first()` and `find_last()` functions I would be OK'ish to allow negative values as seems to be the case for `index-expression`. But in that case, they would not follow slice conventions either, as `index-expression` indices currently do not... Which is a bit silly I think. I am weary to standardize a quirky behaviour, though, and would be more keen to deprecate and forbid negative indices for `index-expression` in the future instead. Which means as imperfect as it is, I would still leave negative indices for `find_` functions currently undefined for now.

Aren't the `find` functions being updated in #119 to exactly align with slices, including not just negative values but also an interpretation of them as relative to the end? I think that actually makes a lot of sense.

> Lastly, `contains()` seems to have a well defined behavior when its first argument is an array, as it is based upon equality with one of the array element and equality is well defined in JMESPath. You are absolutely right, `contains` is not sufficiently defined for string arguments. I think a new JEP is warranted to close the loop and tighten its behavior to only accept a string `search` string and define _containment_ by the intuitive occurence of an _equal_ substring within the subject string. (#121)

Excellent! :+1:

Reply:
> I have actually [updated the JEP-14](https://github.com/jmespath-community/jmespath.spec/blob/main/jep-014-string-functions.md) to require integers for the `$count` parameter to the `split()` and `replace()` functions.

Oh, could you also do the same for `find_first` and `find_last` $start and $end parameters and `pad_left` and `pad_right` $width?

Reply:
> I think a new JEP is warranted to close the loop and tighten its behavior to only accept a string `search` string and define _containment_ by the intuitive occurence of an _equal_ substring within the subject string. (#121)

I've also opened discussions for the other ambiguities:
* out-of-bounds slice indices: https://github.com/jmespath-community/jmespath.spec/discussions/122
* `from_items` input: https://github.com/jmespath-community/jmespath.spec/discussions/123